apps/labs/posts/automated-accessibility-testing-with-bokeh.md

```diff
@@ -0,0 +1,283 @@
+the directories within the Javascript test directory.
+
+We found out that I had been looking at the test setup the wrong way. The bokeh
+test folder has the following [sub
+folders](https://docs.bokeh.org/en/latest/docs/dev_guide/testing.html#select-specific-bokehjs-tests):
+`baselines`, `codebase`, `defaults`, `devtools`, `integration`, and `unit`. The
+`unit`, `defaults`, and `integration` categories of tests were designed to be
+able to both run in the terminal, and on the devtools server. The `unit` and
+`integration` tests looked the most like tests that one would normally see
+and/or write in different projects because they had `describe` blocks and
+`assert` statements. Basing this assumption on just those two facts, I felt the
+location for `playwright` tests would have to model either of those two, to run
+on both terminal & devtools, as well as with assertions. It was on closer
+inspection with my mentor in a pair programming session that we discovered that
+the `codebase` sub-directory used core node modules, and we decided to base
+playwright off that, but imported the necessary parts from `unit` and
+`integration` that would let the assertions and descriptions work as planned.
+This last part had to be done because Bokeh has its own custom-built test
+tooling, similar in syntax to [Jest](https://jestjs.io/), but implemented
+differently. We tested to see if things worked, and they did. We cleaned up the
+code, committed, pushed, and both stood GitHub PR comment section.
+
+On occasion, a question would come up in the PR comment section that would seem
+a bit confusing to me, but then he would clarify what the person was trying to
+ask/communicate, giving me tips on how to implement the suggestions from
+reviewers. It was at that point in time I could really say: "We'd done something
+truly amazing for Bokeh."
+
+## Where we currently are
+
+A testing environment needs tests, and to test what we'd done, we needed a
+variety of valid data visualizations. Luckily, Bokeh very generously had over
+100 example plots in the documentation that we could test against. Testing
+against the documentation came with a few challenges: one of which was the fact
+that the documentation would always need to be built and served locally before
+we could run axe-core tests against it.
+
+Overall, our efforts towards going from 0 to 1 ended up looking like this [PR
+that adds playwright.](https://github.com/bokeh/bokeh/pull/14032)
+
+## Progressive work going forward
+
+Going from 0 to 1 was a good start, but to go from 1 to n, there are more things
+that can be done in Bokeh. In the future, we could:
+
+- Add Integration snapshot tests against regressions and known failures. This
+  would mostly be for accessibility issues found by manual means of testing
+  (rather than through tools), where the problems are of a visual nature. A
+  snapshot could test that future updates do not resurface accessibility
+  problems that have been solved.
```

Comment:
Perhaps I might need to paraphrase it a bit better, but this did come up in one of our weekly syncs. At the time of writing this, it would be in line 387 of our [weekly sync notes](https://hackmd.io/LB1Mx8oUTfeJuevB7R-sDw)